http-client/src/main/java/io/airlift/http/client/GatheringByteArrayInputStream.java

```diff
@@ -0,0 +1,75 @@
+package io.airlift.http.client;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.min;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class GatheringByteArrayInputStream
+        extends InputStream
+{
+    private final Iterator<byte[]> arrays;
```

Comment:
rename these fields to `buffers`, `currentBuffer` and `currentBufferPosition`

Also if you initialize `currentBuffer` and `currentBufferPosition` in the constructor, it will make the code much simpler.  If `buffers` is empty, set them to `byte[0]` and `0`